dask_kubernetes/core.py

```diff
@@ -312,13 +312,62 @@ def scale(self, n):
         KubeCluster.scale_up
         KubeCluster.scale_down
         """
-        pods = self.pods()
+        pods = self._cleanup_succeeded_pods(self.pods())
         if n >= len(pods):
             return self.scale_up(n, pods=pods)
         else:
-            to_close = select_workers_to_close(self.scheduler, len(pods) - n)
+            n_to_delete = len(pods) - n
+            # Before trying to close running workers, check if we can cancel
+            # pending pods (in case the kubernetes cluster was too full to
+            # provision those pods in the first place).
+            running_workers = list(self.scheduler.workers.keys())
+            running_ips = set(urlparse(worker).hostname
+                              for worker in running_workers)
+            pending_pods = [p for p in pods
+                            if p.status.pod_ip not in running_ips]
+            if pending_pods:
+                pending_to_delete = pending_pods[:n_to_delete]
+                logger.debug("Deleting pending pods: %s", pending_to_delete)
+                self._delete_pods(pending_to_delete)
+                n_to_delete = n_to_delete - len(pending_to_delete)
+                if n_to_delete <= 0:
+                    return
+
+            to_close = select_workers_to_close(self.scheduler, n_to_delete)
             logger.debug("Closing workers: %s", to_close)
-            return self.scale_down(to_close)
+            if len(to_close) < len(self.scheduler.workers):
+                # Close workers cleanly to migrate any temporary results to
+                # remaining workers.
+                future = self.scheduler.retire_workers(
+                        workers=to_close, remove=True, close_workers=True)
+
+                def callback(future):
+                    self.scale_down(to_close)
+
+                self.scheduler.loop.add_future(future, callback)
```

Comment:
@mrocklin I have very little experience with tornado / async programming, so I would appreciate your feedback on this pattern.

Reply:
The `scale` method might be called by the user in the IPython interaction thread which is often different from the thread where the IOLoop is running.  As a result we want to avoid calling any coroutine like `retire_workers` directly.  Instead, we need to use the threadsafe method `IOLoop.add_callback`.  I suggest the following:

```python
@gen.coroutine
def f(to_close):
    yield self.scheduler.retire_workers(workers=to_close, ...)
    self.scale_down(to_close)

self.scheduler.loop.add_callback(f, to_close)
```

However this might become problematic if `scale_down` blocks (does it?  I'm not that knowledgable about the kubernetes API).   If so then we might want to offload this to a thread in a thread pool.

```python
from distributed.comm.utils import offload

@gen.coroutine
def f(to_close):
    yield self.scheduler.retire_workers(workers=to_close, ...)
    yield offload(self.scale_down, to_close)

self.scheduler.loop.add_callback(f, to_close)
```